src/runtime-rs/crates/hypervisor/src/firecracker/fc_api.rs

```diff
@@ -0,0 +1,358 @@
+//Copyright (c) 2019-2022 Alibaba Cloud
+//Copyright (c) 2023 Nubificus Ltd
+//
+//SPDX-License-Identifier: Apache-2.0
+
+use crate::firecracker::utils::*;
+use crate::{
+    firecracker::FcInner, kernel_param::KernelParams, NetworkConfig, Param, HYPERVISOR_FIRECRACKER,
+};
+use anyhow::{anyhow, Context, Result};
+use dbs_utils::net::MacAddr;
+use hyper::{Body, Method, Request, Response};
+use hyperlocal::Uri;
+use kata_sys_util::mount;
+use nix::mount::MsFlags;
+use serde_json::json;
+use std::io::ErrorKind;
+use tokio::{fs, fs::File};
+
+const REQUEST_RETRY: u32 = 500;
+const FC_KERNEL: &str = "vmlinux";
+const FC_ROOT_FS: &str = "rootfs";
+const DRIVE_PREFIX: &str = "drive";
+const DISK_POOL_SIZE: u32 = 6;
+
+impl FcInner {
+    pub(crate) fn get_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if self.jailed {
+            self.jail_resource(src, dst)
+        } else {
+            Ok(src.to_string())
+        }
+    }
+
+    fn jail_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if src.is_empty() || dst.is_empty() {
+            return Err(anyhow!("invalid param src {} dst {}", src, dst));
+        }
+
+        let jailed_location = [
+            self.jailer_root.as_str(),
+            HYPERVISOR_FIRECRACKER,
+            &self.id,
+            ROOT,
+            dst,
+        ]
+        .join("/");
+        mount::bind_mount_unchecked(src, jailed_location.as_str(), false, MsFlags::MS_SLAVE)
+            .context("bind_mount ERROR")?;
+
+        let mut abs_path = String::from("/");
+        abs_path.push_str(dst);
+        Ok(abs_path)
+    }
+
+    // Remounting jailer root to ensure it has exec permissions, since firecracker binary will
+    // execute from there
+    pub(crate) async fn remount_jailer_with_exec(&self) -> Result<()> {
+        let dest = &[
+            &self.jailer_root,
+            "/",
+            HYPERVISOR_FIRECRACKER,
+            "/",
+            &self.id,
+            "/",
+            ROOT,
+        ]
+        .concat();
+        let _ = fs::create_dir_all(dest)
+            .await
+            .context(format!("failed to create directory {:?}", dest));
+        mount::bind_mount_unchecked(dest, dest, false, MsFlags::MS_SHARED)
+            .context("bind mount jailer root")?;
+
+        mount::bind_remount(dest, false).context("rebind mount jailer root")?;
+        Ok(())
+    }
+
+    pub(crate) async fn wait_api_socket(&mut self, id: &str) -> Result<()> {
```

Comment:
I'm trying to understand the name of this function - why "wait"?

Reply:
We wait to see if the socket is ready or if something wrong has happened to it, as we cannot proceed if its not

Reply:
ack. we used this for debug purposes (trying to understand whether FC was starting or not, so we were waiting until the socket appeared. Afterwards, we stopped waiting and just used this for stale sockets lying around). Not needed anymore, so we removed that.